<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>One Thousand Plant Transcriptomes Tutorial &mdash; orthoflow 0.1.0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/copybutton.css" type="text/css" />
      <link rel="stylesheet" href="../_static/graphviz.css" type="text/css" />
      <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/clipboard.min.js"></script>
        <script src="../_static/copybutton.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Input" href="input.html" />
    <link rel="prev" title="Reproduction Tutorial" href="reproduction_tutorial.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            orthoflow
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="beginner_tutorial.html">Beginner Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="advanced_tutorial.html">Advanced Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="reproduction_tutorial.html">Reproduction Tutorial</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">One Thousand Plant Transcriptomes Tutorial</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#downloading-and-input-file-creation">Downloading and Input File Creation</a></li>
<li class="toctree-l2"><a class="reference internal" href="#config">Config</a></li>
<li class="toctree-l2"><a class="reference internal" href="#running-orthoflow">Running Orthoflow</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="input.html">Input</a></li>
<li class="toctree-l1"><a class="reference internal" href="workflow.html">Workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="citation.html">Citation and Attribution</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">orthoflow</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">One Thousand Plant Transcriptomes Tutorial</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/main/1kp_tutorial.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<section id="one-thousand-plant-transcriptomes-tutorial">
<h1>One Thousand Plant Transcriptomes Tutorial<a class="headerlink" href="#one-thousand-plant-transcriptomes-tutorial" title="Permalink to this headline"></a></h1>
<p>This tutorial shows you how to run the ‘One Thousand Plant Transcriptomes’ (1KP) dataset in Orthoflow.</p>
<p>This is a dataset of more than 1000 plant transcriptomes for which phylogenomic analysis is presented in:</p>
<blockquote>
<div><p>One Thousand Plant Transcriptomes Initiative. ‘One thousand plant transcriptomes and the phylogenomics of green plants.’ <em>Nature</em> 574, 679–685 (2019). <a class="reference external" href="https://doi.org/10.1038/s41586-019-1693-2">https://doi.org/10.1038/s41586-019-1693-2</a></p>
</div></blockquote>
<section id="downloading-and-input-file-creation">
<h2>Downloading and Input File Creation<a class="headerlink" href="#downloading-and-input-file-creation" title="Permalink to this headline"></a></h2>
<p>You can download the list of species from CyVerse:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>wget<span class="w"> </span>https://de.cyverse.org/anon-files/iplant/home/shared/commons_repo/curated/oneKP_capstone_2019/transcript_assemblies/onekp_SRA.csv
</pre></div>
</div>
<p>That should download a file called <code class="docutils literal notranslate"><span class="pre">onekp_SRA.csv</span></code> to your current directory. The first few lines of the file should look like this:</p>
<div class="highlight-csv notranslate"><div class="highlight"><pre><span></span>OneKP_ID,Accession,Species,NCBI Link
------------,-- -------------,-- ------------------------------,
AALA,ERS1829210,Meliosma cuneifolia,https://www.ncbi.nlm.nih.gov/sra/?term=ERS1829210
ABCD,ERS1830013,Racomitrium elongatum,https://www.ncbi.nlm.nih.gov/sra/?term=ERS1830013
ABEH,ERS1829580,Heliotropium greggii,https://www.ncbi.nlm.nih.gov/sra/?term=ERS1829580
ABIJ,ERS1829938,Selaginella lepidophylla,https://www.ncbi.nlm.nih.gov/sra/?term=ERS1829938
ABSS,ERS1829192,Sassafras albidum,https://www.ncbi.nlm.nih.gov/sra/?term=ERS1829192
ACWS,ERS1829861,Agathis macrophylla,https://www.ncbi.nlm.nih.gov/sra/?term=ERS1829861
ACYX,ERS1829232,Papaver rhoeas,https://www.ncbi.nlm.nih.gov/sra/?term=ERS1829232
ADHK,ERS1829551,Galax urceolata,https://www.ncbi.nlm.nih.gov/sra/?term=ERS1829551
</pre></div>
</div>
<p>To download the transcriptomes and create an input file for Orthoflow, you can use this bash script:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>mkdir<span class="w"> </span>-p<span class="w"> </span>downloads
mkdir<span class="w"> </span>-p<span class="w"> </span>proteins
<span class="nb">echo</span><span class="w"> </span><span class="s2">&quot;file,taxon_string,data_type&quot;</span><span class="w"> </span>&gt;<span class="w"> </span>input_sources_1kp.csv
<span class="k">for</span><span class="w"> </span>DATA<span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="k">$(</span>tail<span class="w"> </span>-n<span class="w"> </span>+3<span class="w"> </span>onekp_SRA.csv<span class="w"> </span><span class="p">|</span><span class="w"> </span>sed<span class="w"> </span><span class="s1">&#39;s/ /_/g&#39;</span><span class="k">)</span><span class="p">;</span><span class="w"> </span><span class="k">do</span>
<span class="w">    </span><span class="nv">ID</span><span class="o">=</span><span class="k">$(</span><span class="nb">echo</span><span class="w"> </span><span class="nv">$DATA</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>cut<span class="w"> </span>-d,<span class="w"> </span>-f1<span class="k">)</span>
<span class="w">    </span><span class="nb">echo</span><span class="w"> </span><span class="nv">$ID</span>

<span class="w">    </span><span class="nv">SPECIES</span><span class="o">=</span><span class="k">$(</span><span class="nb">echo</span><span class="w"> </span><span class="nv">$DATA</span><span class="w"> </span><span class="p">|</span><span class="w"> </span>cut<span class="w"> </span>-d,<span class="w"> </span>-f3<span class="w"> </span><span class="k">)</span>
<span class="w">    </span><span class="nv">ARCHIVE_NAME</span><span class="o">=</span><span class="si">${</span><span class="nv">ID</span><span class="si">}</span>-SOAPdenovo-Trans-translated.tar.bz2
<span class="w">    </span><span class="nv">DOWNLOAD</span><span class="o">=</span>downloads/<span class="nv">$ARCHIVE_NAME</span>
<span class="w">    </span><span class="nv">URL</span><span class="o">=</span>https://de.cyverse.org/anon-files/iplant/home/shared/commons_repo/curated/oneKP_capstone_2019/transcript_assemblies/<span class="si">${</span><span class="nv">ID</span><span class="si">}</span>-<span class="si">${</span><span class="nv">SPECIES</span><span class="si">}</span>/<span class="nv">$ARCHIVE_NAME</span>
<span class="w">    </span><span class="nv">PROTEIN</span><span class="o">=</span>proteins/<span class="nv">$ID</span>.prots.fasta

<span class="w">    </span><span class="c1"># Download if necessary</span>
<span class="w">    </span><span class="o">[[</span><span class="w"> </span>-f<span class="w"> </span><span class="nv">$DOWNLOAD</span><span class="w"> </span><span class="o">]]</span><span class="w"> </span><span class="o">||</span><span class="w"> </span><span class="nb">echo</span><span class="w"> </span><span class="nv">$URL</span>

<span class="w">    </span><span class="k">if</span><span class="w"> </span><span class="o">[</span><span class="w"> </span>-f<span class="w"> </span><span class="nv">$DOWNLOAD</span><span class="w"> </span><span class="o">]</span><span class="w"> </span><span class="p">;</span><span class="w"> </span><span class="k">then</span>
<span class="w">        </span><span class="c1"># Untar archive</span>
<span class="w">        </span><span class="nv">TMP_DIR</span><span class="o">=</span>tmp-<span class="nv">$ID</span>
<span class="w">        </span>mkdir<span class="w"> </span>-p<span class="w"> </span><span class="nv">$TMP_DIR</span>
<span class="w">        </span>tar<span class="w"> </span>xjC<span class="w"> </span><span class="nv">$TMP_DIR</span><span class="w"> </span>-f<span class="w"> </span><span class="nv">$DOWNLOAD</span>

<span class="w">        </span><span class="c1"># Extract the protein file</span>
<span class="w">        </span>mv<span class="w"> </span><span class="nv">$TMP_DIR</span>/*.prots.out<span class="w">  </span><span class="nv">$PROTEIN</span>

<span class="w">        </span><span class="c1"># Clean up the archive data directory</span>
<span class="w">        </span>rm<span class="w"> </span>-rf<span class="w"> </span><span class="nv">$TMP_DIR</span>

<span class="w">        </span><span class="c1"># Save to input file</span>
<span class="w">        </span><span class="nb">echo</span><span class="w"> </span><span class="s2">&quot;</span><span class="nv">$PROTEIN</span><span class="s2">,</span><span class="nv">$SPECIES</span><span class="s2">,Protein&quot;</span><span class="w"> </span>&gt;&gt;<span class="w"> </span>input_sources_1kp.csv
<span class="w">    </span><span class="k">fi</span>
<span class="k">done</span>
</pre></div>
</div>
<p>This will download the available files to the current working directory and it will create an input file for Orthoflow called <code class="docutils literal notranslate"><span class="pre">input_sources_1kp.csv</span></code>.
The first few lines of the file should look like this:</p>
<div class="highlight-csv notranslate"><div class="highlight"><pre><span></span>file,taxon_string,data_type
file,taxon_string,data_type
proteins/AALA.prots.fasta,Meliosma_cuneifolia,Protein
proteins/ABCD.prots.fasta,Racomitrium_elongatum,Protein
proteins/ABEH.prots.fasta,Heliotropium_greggii,Protein
proteins/ABSS.prots.fasta,Sassafras_albidum,Protein
proteins/ACWS.prots.fasta,Agathis_macrophylla,Protein
proteins/AEPI.prots.fasta,Linum_leonii,Protein
proteins/AFLV.prots.fasta,Xerophyllum_asphodeloides,Protein
</pre></div>
</div>
<p>We are using the protein sequences for the transcriptomes, so we set the <code class="docutils literal notranslate"><span class="pre">data_type</span></code> column to <code class="docutils literal notranslate"><span class="pre">Protein</span></code>.</p>
<p>You can now run Orthoflow with this input file.</p>
</section>
<section id="config">
<h2>Config<a class="headerlink" href="#config" title="Permalink to this headline"></a></h2>
<p>Now, let’s set up the configuration file for Orthoflow. We only need to set the parameters which are different from the default values.</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">input_sources</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;input_sources_1kp.csv&quot;</span><span class="w"> </span><span class="c1"># Alternatively this can be specified on the command line with --files input_sources_1kp.csv</span>
<span class="nt">infer_tree_with_cds_seqs</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span><span class="w"> </span><span class="c1"># We are using protein sequences</span>
<span class="nt">supermatrix</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span><span class="w">  </span><span class="c1"># We will only infer a tree using the supertree (ASTRAL) method</span>
</pre></div>
</div>
<p>Save this file to <code class="docutils literal notranslate"><span class="pre">config_1kp.yaml</span></code>.</p>
</section>
<section id="running-orthoflow">
<h2>Running Orthoflow<a class="headerlink" href="#running-orthoflow" title="Permalink to this headline"></a></h2>
<p>Phylogenomic analysis of this dataset can then be run with the command:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>orthoflow<span class="w"> </span>--configfile<span class="w"> </span>config_1kp.yaml
</pre></div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="reproduction_tutorial.html" class="btn btn-neutral float-left" title="Reproduction Tutorial" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="input.html" class="btn btn-neutral float-right" title="Input" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022 University of Melbourne.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>