<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Beginner Tutorial &mdash; orthoflow 0.1.0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/copybutton.css" type="text/css" />
      <link rel="stylesheet" href="../_static/graphviz.css" type="text/css" />
      <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/clipboard.min.js"></script>
        <script src="../_static/copybutton.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Advanced Tutorial" href="advanced_tutorial.html" />
    <link rel="prev" title="Installation" href="installation.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            orthoflow
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Beginner Tutorial</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#input-data">Input data</a></li>
<li class="toctree-l2"><a class="reference internal" href="#simple-run">Simple run</a></li>
<li class="toctree-l2"><a class="reference internal" href="#examining-the-output">Examining the output</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#inferred-tree-and-intermediate-files">Inferred tree and intermediate files</a></li>
<li class="toctree-l3"><a class="reference internal" href="#report-and-diagnostics">Report and diagnostics</a></li>
<li class="toctree-l3"><a class="reference internal" href="#output-logs">Output logs</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="advanced_tutorial.html">Advanced Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="reproduction_tutorial.html">Reproduction Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="1kp_tutorial.html">One Thousand Plant Transcriptomes Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="input.html">Input</a></li>
<li class="toctree-l1"><a class="reference internal" href="workflow.html">Workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="citation.html">Citation and Attribution</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">orthoflow</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Beginner Tutorial</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/main/beginner_tutorial.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<section id="beginner-tutorial">
<h1>Beginner Tutorial<a class="headerlink" href="#beginner-tutorial" title="Permalink to this headline"></a></h1>
<section id="input-data">
<h2>Input data<a class="headerlink" href="#input-data" title="Permalink to this headline"></a></h2>
<p>Orthoflow works from an input CSV file with information about the data sources  to be used. Preparing this file is central to setting up your run. The default filename for this is <code class="docutils literal notranslate"><span class="pre">input_sources.csv</span></code>.</p>
<p>It needs the columns <code class="docutils literal notranslate"><span class="pre">file</span></code>, <code class="docutils literal notranslate"><span class="pre">taxon_string</span></code>, <code class="docutils literal notranslate"><span class="pre">data_type</span></code> and <code class="docutils literal notranslate"><span class="pre">translation_table</span></code>.</p>
<ul class="simple">
<li><p>The <code class="docutils literal notranslate"><span class="pre">file</span></code> column is the path to the file relative to the working directory.</p></li>
<li><p>The <code class="docutils literal notranslate"><span class="pre">taxon_string</span></code> is the name of the taxon from which the data was obtained.</p></li>
<li><p>The <code class="docutils literal notranslate"><span class="pre">data_type</span></code> column should be <code class="docutils literal notranslate"><span class="pre">GenBank</span></code> when providing a GenBank-formatted file with CDS annotations, or <code class="docutils literal notranslate"><span class="pre">CDS</span></code> or <code class="docutils literal notranslate"><span class="pre">Protein</span></code> when providing a FASTA file with coding sequences consisting of nucleotides or amino acids respectively.</p></li>
<li><p>The <code class="docutils literal notranslate"><span class="pre">translation_table</span></code> column should have the translation table (genetic code) number for the data as given <a class="reference external" href="https://www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi?mode=c">here</a>.</p></li>
</ul>
<p>Let’s look at the demonstration dataset distributed with the code: <code class="docutils literal notranslate"><span class="pre">tests/test-data/input_sources.csv</span></code>.</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 24%" />
<col style="width: 43%" />
<col style="width: 13%" />
<col style="width: 21%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>file</p></th>
<th class="head"><p>taxon_string</p></th>
<th class="head"><p>data_type</p></th>
<th class="head"><p>translation_table</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>KY509313.gb</p></td>
<td><p>Avrainvillea_mazei_HV02664</p></td>
<td><p>GenBank</p></td>
<td><p>11</p></td>
</tr>
<tr class="row-odd"><td><p>NC_026795.txt</p></td>
<td><p>Bryopsis_plumosa_WEST4718</p></td>
<td><p>GenBank</p></td>
<td><p>11</p></td>
</tr>
<tr class="row-even"><td><p>KX808498.gb</p></td>
<td><p>Caulerpa_cliftonii_HV03798</p></td>
<td><p>GenBank</p></td>
<td><p>11</p></td>
</tr>
<tr class="row-odd"><td><p>KY819064.cds.fasta</p></td>
<td><p>Chlorodesmis_fastigiata_HV03865</p></td>
<td><p>CDS</p></td>
<td><p>11</p></td>
</tr>
<tr class="row-even"><td><p>KX808497.fa</p></td>
<td><p>Derbesia_sp_WEST4838</p></td>
<td><p>CDS</p></td>
<td><p>11</p></td>
</tr>
<tr class="row-odd"><td><p>MH591079.gb</p></td>
<td><p>Dichotomosiphon_tuberosus_HV03781</p></td>
<td><p>GenBank</p></td>
<td><p>11</p></td>
</tr>
<tr class="row-even"><td><p>MH591080.gbk</p></td>
<td><p>Dichotomosiphon_tuberosus_HV03781</p></td>
<td><p>GenBank</p></td>
<td><p>11</p></td>
</tr>
<tr class="row-odd"><td><p>MH591081.gbk</p></td>
<td><p>Dichotomosiphon_tuberosus_HV03781</p></td>
<td><p>GenBank</p></td>
<td><p>11</p></td>
</tr>
<tr class="row-even"><td><p>MH591083.gb</p></td>
<td><p>Flabellia_petiolata_HV01202</p></td>
<td><p>GenBank</p></td>
<td><p>11</p></td>
</tr>
<tr class="row-odd"><td><p>MH591084.gb</p></td>
<td><p>Flabellia_petiolata_HV01202</p></td>
<td><p>GenBank</p></td>
<td><p>11</p></td>
</tr>
<tr class="row-even"><td><p>MH591085.gb</p></td>
<td><p>Flabellia_petiolata_HV01202</p></td>
<td><p>GenBank</p></td>
<td><p>11</p></td>
</tr>
<tr class="row-odd"><td><p>MH591086.gb</p></td>
<td><p>Flabellia_petiolata_HV01202</p></td>
<td><p>GenBank</p></td>
<td><p>11</p></td>
</tr>
</tbody>
</table>
<p>We are using a dataset of algal chloroplast genomes, some as annotated genbank files (<code class="docutils literal notranslate"><span class="pre">data_type:</span> <span class="pre">Genbank</span></code>), some as fasta files of the coding sequences (<code class="docutils literal notranslate"><span class="pre">data_type:</span> <span class="pre">CDS</span></code>). They all use the bacterial genetic code (<code class="docutils literal notranslate"><span class="pre">translation_table:</span> <span class="pre">11</span></code>). Some of the genomes were in a single Genbank file (e.g. <code class="docutils literal notranslate"><span class="pre">KY09313.gb</span></code> at the top), others were fragmented across multiple files (e.g. last 4 all belonging to the same taxon).</p>
<p>The <code class="docutils literal notranslate"><span class="pre">taxon_string</span></code> column is perhaps the most important one, as these will be the names to appear in the output tree and this determines how input data gets grouped (e.g. all CDSs in the final four GenBank files will be grouped into a single taxon). In this case, we have included specimen numbers as part of the taxon string but that is optional.</p>
</section>
<section id="simple-run">
<h2>Simple run<a class="headerlink" href="#simple-run" title="Permalink to this headline"></a></h2>
<p>We are using the small demonstration dataset distributed with the Orthoflow in the <code class="docutils literal notranslate"><span class="pre">tests/test-data</span></code> subdirectory.</p>
<p>Go into the directory containing the <code class="docutils literal notranslate"><span class="pre">input_sources.csv</span></code> file and run orthoflow with default settings with these commands:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">cd</span> <span class="n">tests</span><span class="o">/</span><span class="n">test</span><span class="o">-</span><span class="n">data</span>
<span class="n">orthoflow</span>
</pre></div>
</div>
<p>By default, Orthoflow will extract the CDSs from the input files, run OrthoFinder followed by OrthoSNAP to determine orthologous genes, align them and infer a concatenated tree from the protein sequences. You can follow progress on the screen as the workflow executes and outputs are produced.</p>
<p>Note that the first time you run the workflow, it will be slow because it needs to download and install the software it depends on. This is a one-time thing and runs should get going much faster after.</p>
</section>
<section id="examining-the-output">
<h2>Examining the output<a class="headerlink" href="#examining-the-output" title="Permalink to this headline"></a></h2>
<section id="inferred-tree-and-intermediate-files">
<h3>Inferred tree and intermediate files<a class="headerlink" href="#inferred-tree-and-intermediate-files" title="Permalink to this headline"></a></h3>
<p>All output files are saved in the <code class="docutils literal notranslate"><span class="pre">results</span></code> directory. Output files are subdivided into the workflow modules, which each have their own subdirectory. For the demonstration analysis that we ran above, the inferred phylogeny will be in the <code class="docutils literal notranslate"><span class="pre">supermatrix</span></code> subdirectory and be called <code class="docutils literal notranslate"><span class="pre">supermatrix.protein.fa.treefile</span></code>. Open this with a tree browser (e.g. <a class="reference external" href="https://github.com/rambaut/figtree">FigTree</a>). Also take some time to browse the intermediary files, including the orthogroups, gene alignments and the supermatrix constructed from them.</p>
</section>
<section id="report-and-diagnostics">
<h3>Report and diagnostics<a class="headerlink" href="#report-and-diagnostics" title="Permalink to this headline"></a></h3>
<p>The report provides an overview of the results, the analysis settings used and citations of the software used to produce the results. This report is found in the <code class="docutils literal notranslate"><span class="pre">results/report.cds.html</span></code> and/or <code class="docutils literal notranslate"><span class="pre">results/report.protein.html</span></code>, depending on the method used to infer the phylogeny.</p>
</section>
<section id="output-logs">
<h3>Output logs<a class="headerlink" href="#output-logs" title="Permalink to this headline"></a></h3>
<p>The output logs of all software used as part of the workflow can be found in the <code class="docutils literal notranslate"><span class="pre">logs</span></code> directory.</p>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="installation.html" class="btn btn-neutral float-left" title="Installation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="advanced_tutorial.html" class="btn btn-neutral float-right" title="Advanced Tutorial" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022 University of Melbourne.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>