<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Citation and Attribution &mdash; orthoflow 0.1.0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/copybutton.css" type="text/css" />
      <link rel="stylesheet" href="../_static/graphviz.css" type="text/css" />
      <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/clipboard.min.js"></script>
        <script src="../_static/copybutton.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="prev" title="Contributing" href="contributing.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="../index.html" class="icon icon-home"> orthoflow
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="beginner_tutorial.html">Beginner Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="beginner_tutorial.html#quick-start-guide">Quick start guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="advanced_tutorial.html">Advanced Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="input.html">Input</a></li>
<li class="toctree-l1"><a class="reference internal" href="workflow.html">Workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Citation and Attribution</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#bibliography">Bibliography</a></li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">orthoflow</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home"></a></li>
      <li class="breadcrumb-item active">Citation and Attribution</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/main/citation.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<section id="citation-and-attribution">
<h1>Citation and Attribution<a class="headerlink" href="#citation-and-attribution" title="Permalink to this headline"></a></h1>
<p>Orthoflow was created by Robert Turnbull, Jacob Steenwyk, Simon Mutch, Vinícius Salazar, and Heroen Verbruggen.</p>
<p>Citation details to follow.</p>
<section id="bibliography">
<h2>Bibliography<a class="headerlink" href="#bibliography" title="Permalink to this headline"></a></h2>
<p>The following articles and software packages were used in this workflow:</p>
<div class="docutils container" id="id1">
<dl class="citation">
<dt class="label" id="id3"><span class="brackets">CAC+09</span></dt>
<dd><p>Peter J. A. Cock, Tiago Antao, Jeffrey T. Chang, Brad A. Chapman, Cymon J. Cox, Andrew Dalke, Iddo Friedberg, Thomas Hamelryck, Frank Kauff, Bartek Wilczynski, and Michiel J. L. de Hoon. Biopython: freely available Python tools for computational molecular biology and bioinformatics. <em>Bioinformatics</em>, 25(11):1422–1423, 03 2009. URL: <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btp163">https://doi.org/10.1093/bioinformatics/btp163</a>, <a class="reference external" href="https://arxiv.org/abs/https://academic.oup.com/bioinformatics/article-pdf/25/11/1422/944180/btp163.pdf">arXiv:https://academic.oup.com/bioinformatics/article-pdf/25/11/1422/944180/btp163.pdf</a>, <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btp163">doi:10.1093/bioinformatics/btp163</a>.</p>
</dd>
<dt class="label" id="id4"><span class="brackets">Eat20</span></dt>
<dd><p>Deren A. R. Eaton. Toytree: A minimalist tree visualization and manipulation library for Python. <em>Methods in Ecology and Evolution</em>, 11:187–191, 2020. <a class="reference external" href="https://doi.org/10.1111/2041-210X.13313">doi:10.1111/2041-210X.13313</a>.</p>
</dd>
<dt class="label" id="id6"><span class="brackets">EK19</span></dt>
<dd><p>David M. Emms and Steven Kelly. Orthofinder: phylogenetic orthology inference for comparative genomics. <em>Genome Biology</em>, 20(1):238, 2019. URL: <a class="reference external" href="https://doi.org/10.1186/s13059-019-1832-y">https://doi.org/10.1186/s13059-019-1832-y</a>, <a class="reference external" href="https://doi.org/10.1186/s13059-019-1832-y">doi:10.1186/s13059-019-1832-y</a>.</p>
</dd>
<dt class="label" id="id8"><span class="brackets">HCvH+17</span></dt>
<dd><p>Diep Thi Hoang, Olga Chernomor, Arndt von Haeseler, Bui Quang Minh, and Le Sy Vinh. UFBoot2: Improving the Ultrafast Bootstrap Approximation. <em>Molecular Biology and Evolution</em>, 35(2):518–522, 10 2017. URL: <a class="reference external" href="https://doi.org/10.1093/molbev/msx281">https://doi.org/10.1093/molbev/msx281</a>, <a class="reference external" href="https://arxiv.org/abs/https://academic.oup.com/mbe/article-pdf/35/2/518/24367824/msx281.pdf">arXiv:https://academic.oup.com/mbe/article-pdf/35/2/518/24367824/msx281.pdf</a>, <a class="reference external" href="https://doi.org/10.1093/molbev/msx281">doi:10.1093/molbev/msx281</a>.</p>
</dd>
<dt class="label" id="id10"><span class="brackets">KMW+17</span></dt>
<dd><p>Subha Kalyaanamoorthy, Bui Quang Minh, Thomas K F Wong, Arndt von Haeseler, and Lars S Jermiin. Modelfinder: fast model selection for accurate phylogenetic estimates. <em>Nature Methods</em>, 14(6):587–589, 2017. URL: <a class="reference external" href="https://doi.org/10.1038/nmeth.4285">https://doi.org/10.1038/nmeth.4285</a>, <a class="reference external" href="https://doi.org/10.1038/nmeth.4285">doi:10.1038/nmeth.4285</a>.</p>
</dd>
<dt class="label" id="id9"><span class="brackets">MSC+20</span></dt>
<dd><p>Bui Quang Minh, Heiko A Schmidt, Olga Chernomor, Dominik Schrempf, Michael D Woodhams, Arndt von Haeseler, and Robert Lanfear. IQ-TREE 2: New Models and Efficient Methods for Phylogenetic Inference in the Genomic Era. <em>Molecular Biology and Evolution</em>, 37(5):1530–1534, 02 2020. URL: <a class="reference external" href="https://doi.org/10.1093/molbev/msaa015">https://doi.org/10.1093/molbev/msaa015</a>, <a class="reference external" href="https://arxiv.org/abs/https://academic.oup.com/mbe/article-pdf/37/5/1530/33386032/msaa015.pdf">arXiv:https://academic.oup.com/mbe/article-pdf/37/5/1530/33386032/msaa015.pdf</a>, <a class="reference external" href="https://doi.org/10.1093/molbev/msaa015">doi:10.1093/molbev/msaa015</a>.</p>
</dd>
<dt class="label" id="id7"><span class="brackets">SBL+21</span></dt>
<dd><p>Jacob L Steenwyk, III Buida, Thomas J, Abigail L Labella, Yuanning Li, Xing-Xing Shen, and Antonis Rokas. PhyKIT: a broadly applicable UNIX shell toolkit for processing and analyzing phylogenomic data. <em>Bioinformatics</em>, 37(16):2325–2331, 02 2021. URL: <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btab096">https://doi.org/10.1093/bioinformatics/btab096</a>, <a class="reference external" href="https://arxiv.org/abs/https://academic.oup.com/bioinformatics/article-pdf/37/16/2325/39948152/btab096.pdf">arXiv:https://academic.oup.com/bioinformatics/article-pdf/37/16/2325/39948152/btab096.pdf</a>, <a class="reference external" href="https://doi.org/10.1093/bioinformatics/btab096">doi:10.1093/bioinformatics/btab096</a>.</p>
</dd>
<dt class="label" id="id2"><span class="brackets">SBGonccalves+21</span></dt>
<dd><p>Jacob L. Steenwyk, Thomas J. Buida, Carla Gonçalves, Dayna C. Goltz, Grace Morales, Matthew E. Mead, Abigail L. LaBella, Christina M. Chavez, Jonathan E. Schmitz, Maria Hadjifrangiskou, Yuanning Li, and Antonis Rokas. BioKIT: a versatile toolkit for processing and analyzing diverse types of sequence data. <em>biorxiv</em>, oct 2021. URL: <a class="reference external" href="https://doi.org/10.1101\%2F2021.10.02.462868">https://doi.org/10.1101\%2F2021.10.02.462868</a>, <a class="reference external" href="https://doi.org/10.1101/2021.10.02.462868">doi:10.1101/2021.10.02.462868</a>.</p>
</dd>
<dt class="label" id="id5"><span class="brackets">ZRSM18</span></dt>
<dd><p>Chao Zhang, Maryam Rabiee, Erfan Sayyari, and Siavash Mirarab. Astral-iii: polynomial time species tree reconstruction from partially resolved gene trees. <em>BMC Bioinformatics</em>, 19(6):153, 2018. URL: <a class="reference external" href="https://doi.org/10.1186/s12859-018-2129-y">https://doi.org/10.1186/s12859-018-2129-y</a>, <a class="reference external" href="https://doi.org/10.1186/s12859-018-2129-y">doi:10.1186/s12859-018-2129-y</a>.</p>
</dd>
</dl>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="contributing.html" class="btn btn-neutral float-left" title="Contributing" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022 University of Melbourne.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>