<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Workflow development guidelines &mdash; orthoflow 0.1.0 documentation</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/copybutton.css" type="text/css" />
      <link rel="stylesheet" href="../_static/graphviz.css" type="text/css" />
      <link rel="stylesheet" href="../_static/custom.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/jquery.js"></script>
        <script src="../_static/underscore.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/clipboard.min.js"></script>
        <script src="../_static/copybutton.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            orthoflow
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="beginner_tutorial.html">Beginner Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="beginner_tutorial.html#quick-start-guide">Quick start guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="advanced_tutorial.html">Advanced Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="input.html">Input</a></li>
<li class="toctree-l1"><a class="reference internal" href="workflow.html">Workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="citation.html">Citation and Attribution</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">orthoflow</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Workflow development guidelines</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/main/workflow_dev.md.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<section id="workflow-development-guidelines">
<h1>Workflow development guidelines<a class="headerlink" href="#workflow-development-guidelines" title="Permalink to this headline"></a></h1>
<section id="diagram">
<h2>Diagram<a class="headerlink" href="#diagram" title="Permalink to this headline"></a></h2>
<p>Link to original on <a class="reference external" href="https://app.diagrams.net/#G1feeWJLnqI-_EN4x8zY2V9GChWosKFZg3">diagrams.net</a></p>
<p><img alt="" src="_static/orthoflow-workflow-diagram.svg" /></p>
</section>
<section id="user-input">
<h2>User input<a class="headerlink" href="#user-input" title="Permalink to this headline"></a></h2>
<p>Brain dump of what we need for the user to provide other than a directory with data files. Definitely incomplete so please add your thoughts.</p>
<ul class="simple">
<li><p>flow control</p>
<ul>
<li><p>use SC-OG only, SNAP-OG only, or both combined</p></li>
<li><p>type of ortholog detection, with checks for adequate input (i.e. reference database for fishing)</p></li>
<li><p>run trees from AA or NT, I suggest not allowing both for the same run as that also affects other settings (e.g. model types)</p></li>
<li><p>whole series of alignment filtering options</p></li>
</ul>
</li>
<li><p>program settings</p>
<ul>
<li><p>IQtree: model (with option to carry out model selection): could use IQtree model definition syntax</p></li>
<li><p>IQtree: type of branch support</p></li>
</ul>
</li>
</ul>
</section>
<section id="data-intake-module">
<h2>Data intake module<a class="headerlink" href="#data-intake-module" title="Permalink to this headline"></a></h2>
<p>Must work from input CSV file to determine which type of input the different input files represent. It produces two files for each taxon, one with CDS and one with protein sequences (translated from CDS based on genetic code).</p>
<p>The default name for the input CSV file is <code class="docutils literal notranslate"><span class="pre">input_sources.csv</span></code>. To override that, either edit the <code class="docutils literal notranslate"><span class="pre">config.yml</span></code> file or use pass an argument to the CLI overriding the config setting, e.g. <code class="docutils literal notranslate"><span class="pre">--config</span> <span class="pre">input_sources='/path/to/desired/input/file.csv'</span></code> (see <a class="reference external" href="https://snakemake.readthedocs.io/en/stable/executing/cli.html">here</a> for more details).</p>
<ol class="arabic simple">
<li><p>Intake of exome data</p>
<ul class="simple">
<li><p>this comes in fasta format</p></li>
<li><p>add taxon_string field from input CSV with the <code class="docutils literal notranslate"><span class="pre">add_taxon_to_seqID.py</span></code> in <code class="docutils literal notranslate"><span class="pre">ext_scripts</span></code>, keep this output</p></li>
<li><p>translate sequences using <code class="docutils literal notranslate"><span class="pre">biokit</span> <span class="pre">translate</span></code> (<a class="reference external" href="https://github.com/JLSteenwyk/BioKIT">github</a>), specifying the genetic code given in the input CSV, keep output</p></li>
</ul>
</li>
<li><p>Intake of annotated genbank files</p>
<ul class="simple">
<li><p>this comes in genbank flatfile format</p></li>
<li><p>extract CDS features from genbank file with <code class="docutils literal notranslate"><span class="pre">gbseqextractor</span></code> (<a class="reference external" href="https://github.com/linzhi2013/gbseqextractor">github</a>)</p></li>
<li><p>this essentially converts the input into the exome data from point 1, so downstream processing is identical to 1 from above</p></li>
<li><p>add taxon_string field from <code class="docutils literal notranslate"><span class="pre">input\_sources.csv</span></code>, keep this output</p></li>
<li><p>translate sequences using <code class="docutils literal notranslate"><span class="pre">biokit</span> <span class="pre">translate</span></code>, specifying genetic code given in input CSV, keep output
2b.  Index CDSs</p></li>
<li><p>create samtools faidx indexed copy of all CDS combined (using files with taxon string added but before translation)</p></li>
<li><p>first all the right fasta files need to be concatenated: <code class="docutils literal notranslate"><span class="pre">cat</span> <span class="pre">*.fa</span> <span class="pre">&gt;</span> <span class="pre">all_CDS.fa</span></code></p></li>
<li><p>next this needs to be fed into samtools for indexing: <code class="docutils literal notranslate"><span class="pre">samtools</span> <span class="pre">faidx</span> <span class="pre">--fai-idx</span> <span class="pre">all_CDS.fai</span> <span class="pre">all_CDS.fa</span></code></p></li>
<li><p>note: it may be possible to pipe the cat output directly into the samtools command and save a step (and a potentially huge intermediate file that no-one would need for other purposes)</p></li>
</ul>
</li>
</ol>
</section>
<section id="ortholog-module-de-novo-stream">
<h2>Ortholog module, de novo stream<a class="headerlink" href="#ortholog-module-de-novo-stream" title="Permalink to this headline"></a></h2>
<p>Uses the translated sequences from the data intake module, clusters them into orthogroups (gene clusters) based on pairwise similarity, then filters these down to the single copy orthogroups (SC-OG) and SNAP orthogroups (SNAP-OG). This module also includes a filtering step for occupancy, i.e. how many taxa are present in the inferred orthogroup.</p>
<ol class="arabic simple" start="3">
<li><p>Orthology inference</p>
<ul class="simple">
<li><p>based on the translated CDSs, run an orthology inference program</p></li>
<li><p>at this stage, I would like two algorithms to be implemented here, OrthoFinder and OrthoMCL</p></li>
<li><p>this step is where I’d like to implement some extra alternatives later for comparative analysis</p></li>
<li><p>the output differs between programs, but for orthofinder, among the many outputs are the orthogroup sequences (fasta files) and phylogenetic trees (one per orthogroup)</p></li>
<li><p>There are too many OGs that are not phylogenetically informative (too few sequences in them – for nuclear genes many are singletons). I wrote a script <code class="docutils literal notranslate"><span class="pre">filter_OrthoFinder.py</span></code> that runs through the output directory of an OrthoFinder run and copies the OGs with more than a user-defined number of sequences to a new directory to reduce the burden of downstream processing. Ideally the minimum number of sequences is set to the occupancy threshold defined by the user.</p></li>
<li><p>haven’t used OrthoMCL recently, so I’m not sure about the outputs it produces, so let’s put a placeholder in the workflow but implement this later</p></li>
</ul>
</li>
<li><p>Filtering SC-OG</p>
<ul class="simple">
<li><p>from the output of step 3, filter out the single copy orthogroups</p></li>
<li><p>I wrote the <code class="docutils literal notranslate"><span class="pre">filter_SCOG.pl</span></code> script to do this. If the data input into this script are prepared with <code class="docutils literal notranslate"><span class="pre">filter_OrthoFinder.py</span></code> then it already meets the occupancy threshold given by the user.</p></li>
</ul>
</li>
<li><p>Inferring and SNAP-OG</p>
<ul class="simple">
<li><p>make this step optional, as not all users will want to use this</p></li>
<li><p>run the OGs that are not SC-OGs through OrthoSNAP</p></li>
<li><p>use the occupancy threshold given by the user</p></li>
<li><p>Note: At the moment SNAP-OG works from the trees inferred in OrthoFinder. Ideally we’d implement our own tree inference including support values as that allows OrthoSNAP to take phylogenetic uncertainty into account.</p></li>
</ul>
</li>
<li><p>Post-orthogroup organisation</p>
<ul class="simple">
<li><p>produce clean output for subsequent module</p></li>
<li><p>retain only taxon string in sequence ID</p></li>
<li><p>only protein sequences are used in ortholog module, but also need corresponding CDS</p></li>
<li><p>use filenames that indicate they are SC or SNAP</p></li>
<li><p>retain output files, important intermediary output</p></li>
</ul>
</li>
</ol>
</section>
<section id="alignment-module">
<h2>Alignment module<a class="headerlink" href="#alignment-module" title="Permalink to this headline"></a></h2>
<p>Uses the cleaned-up OG files from 7, where we have corresponding CDS and protein files. I denoted these NT (for CDS) and AA (amino acid, i.e. protein). In this module, the aim is to align the sequences and do quality filtering.</p>
<ol class="arabic simple" start="7">
<li><p>Alignment</p>
<ul class="simple">
<li><p>first align the protein (amino acid) file with MAFFT</p></li>
<li><p>Then back-translate the alignment to codons based on the CDS sequences, yielding a correspond alignment of nucleotide sequences.</p></li>
<li><p>I wrote a snakemake rule for the MAFFT step: mafft_aa.</p></li>
<li><p>This will need to be updated to point to the right input and output directories</p></li>
</ul>
</li>
<li><p>Filtering</p>
<ul class="simple">
<li><p>To be considered an optional step.</p></li>
<li><p>We talked about this during our 8 April meeting. Steps to be specified in further detail.</p></li>
</ul>
</li>
</ol>
<p>8b. Wrapping up alignment module</p>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>- Aligned (amino acid) sequences need to be translated back to the original CDS. Two steps need to be taken to achieve this. First, the original CDSs need to be located. I wrote a prototype script (matching_CDSs.py) that does this. This needs to be updated to point to the right input files. The second step is to create a codon-based alignment of the CDSs that matches the AA alignment. Phykit implements this, with the thread_dna function.
- At the end, the sequence IDs need to be trimmed down to contain just the taxon identifier and produce clean output for the next stages. I wrote &quot;ext_scripts/seqID_taxon_only.pl&quot; to do this and added a rule for this. We probably want this running in a conda environment to be safe but I haven&#39;t done that. 
- Retain output files of both AA and NT, important intermediary output
</pre></div>
</div>
</section>
<section id="gene-tree-module">
<h2>Gene tree module<a class="headerlink" href="#gene-tree-module" title="Permalink to this headline"></a></h2>
<p>This is an important module that serves multiple purposes. One is to produce the final gene trees, but the module will probably also need to be plugged in to produce intermediary trees as part of the orthologs module, so we need to build this with flexibility in mind. The user will need to specify whether they want trees to be inferred from NT or AA data, and this flexibility needs to be programmed into this somehow. Not sure if this flexibility sits within the module itself, as that may limit the utility of this module for other purposes.</p>
<p>Steps TBD</p>
</section>
<section id="reconciliation-module">
<h2>Reconciliation module<a class="headerlink" href="#reconciliation-module" title="Permalink to this headline"></a></h2>
<p>TBD</p>
</section>
<section id="supermatrix-module">
<h2>Supermatrix module<a class="headerlink" href="#supermatrix-module" title="Permalink to this headline"></a></h2>
<ol class="arabic simple" start="11">
<li><p>PhyKIT</p>
<ul class="simple">
<li><p>a file needs to be produced listing all alignment files that need to be concatenated, say this is called files.txt – I wrote a rule for this but doubt this meets proper workflow development standards</p></li>
<li><p>Heroen created a conda environment for PhyKIT (envs/phykit.yaml)</p></li>
<li><p>Heroen wrote a concatenation rule that uses Phykit to create the supermatrix.</p></li>
</ul>
</li>
<li><p>IQtree phylogenetic inference</p>
<ul class="simple">
<li><p>Heroen added an iqtree environment and wrote a simple iqtree command that will do the job for now</p></li>
<li><p>Will need updating to include user-specified parameters (at minimum: specification of model or model selection and bootstrapping params) once we know</p></li>
</ul>
</li>
</ol>
</section>
<section id="reporting-module">
<h2>Reporting module<a class="headerlink" href="#reporting-module" title="Permalink to this headline"></a></h2>
<p>This isn’t illustrated on the workflow, but we need to provide a clean report to the user with final results, intermediate outputs, program settings etc.</p>
<p>I feel very strongly that this also includes a list of all the papers users should cite for appropriate acknowledgement of the software that is being used under the hood. This will differ depending on the run, but should be easy to manage.</p>
</section>
</section>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022 University of Melbourne.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>