<dl>
    <dt>1</dt>
    <dd>Deren A.&nbsp;R. Eaton.
        Toytree: <span class="bibtex-protected">A</span> minimalist tree visualization and manipulation library for
        <span class="bibtex-protected">Python</span>.
        <em>Methods in Ecology and Evolution</em>, 11:187–191, 2020.
        <a href="https://doi.org/10.1111/2041-210X.13313">doi:10.1111/2041-210X.13313</a>.
    </dd>
    <dt>2</dt>
    <dd>David&nbsp;M. Emms and Steven Kelly.
        Orthofinder: phylogenetic orthology inference for comparative genomics.
        <em>Genome Biology</em>, 20(1):238, 2019.
        URL: <a href="https://doi.org/10.1186/s13059-019-1832-y">https://doi.org/10.1186/s13059-019-1832-y</a>, <a
            href="https://doi.org/10.1186/s13059-019-1832-y">doi:10.1186/s13059-019-1832-y</a>.
    </dd>
    <dt>3</dt>
    <dd>Diep&nbsp;Thi Hoang, Olga Chernomor, Arndt von Haeseler, Bui&nbsp;Quang Minh, and Le&nbsp;Sy Vinh.
        <span class="bibtex-protected">UFBoot2: Improving the Ultrafast Bootstrap Approximation</span>.
        <em>Molecular Biology and Evolution</em>, 35(2):518&ndash;522, 10 2017.
        URL: <a href="https://doi.org/10.1093/molbev/msx281">https://doi.org/10.1093/molbev/msx281</a>, <a
            href="https://arxiv.org/abs/https://academic.oup.com/mbe/article-pdf/35/2/518/24367824/msx281.pdf">arXiv:https://academic.oup.com/mbe/article-pdf/35/2/518/24367824/msx281.pdf</a>,
        <a href="https://doi.org/10.1093/molbev/msx281">doi:10.1093/molbev/msx281</a>.
    </dd>
    <dt>4</dt>
    <dd>Subha Kalyaanamoorthy, Bui&nbsp;Quang Minh, Thomas K&nbsp;F Wong, Arndt von Haeseler, and Lars&nbsp;S Jermiin.
        Modelfinder: fast model selection for accurate phylogenetic estimates.
        <em>Nature Methods</em>, 14(6):587–589, 2017.
        URL: <a href="https://doi.org/10.1038/nmeth.4285">https://doi.org/10.1038/nmeth.4285</a>, <a
            href="https://doi.org/10.1038/nmeth.4285">doi:10.1038/nmeth.4285</a>.
    </dd>
    <dt>5</dt>
    <dd>Kazutaka Katoh and Daron&nbsp;M. Standley.
        <span class="bibtex-protected">MAFFT Multiple Sequence Alignment Software Version 7: Improvements in Performance
            and Usability</span>.
        <em>Molecular Biology and Evolution</em>, 30(4):772&ndash;780, 01 2013.
        URL: <a href="https://doi.org/10.1093/molbev/mst010">https://doi.org/10.1093/molbev/mst010</a>, <a
            href="https://arxiv.org/abs/https://academic.oup.com/mbe/article-pdf/30/4/772/6420419/mst010.pdf">arXiv:https://academic.oup.com/mbe/article-pdf/30/4/772/6420419/mst010.pdf</a>,
        <a href="https://doi.org/10.1093/molbev/mst010">doi:10.1093/molbev/mst010</a>.
    </dd>
    <dt>6</dt>
    <dd>Bui&nbsp;Quang Minh, Heiko&nbsp;A Schmidt, Olga Chernomor, Dominik Schrempf, Michael&nbsp;D Woodhams, Arndt von
        Haeseler, and Robert Lanfear.
        <span class="bibtex-protected">IQ-TREE 2: New Models and Efficient Methods for Phylogenetic Inference in the
            Genomic Era</span>.
        <em>Molecular Biology and Evolution</em>, 37(5):1530&ndash;1534, 02 2020.
        URL: <a href="https://doi.org/10.1093/molbev/msaa015">https://doi.org/10.1093/molbev/msaa015</a>, <a
            href="https://arxiv.org/abs/https://academic.oup.com/mbe/article-pdf/37/5/1530/33386032/msaa015.pdf">arXiv:https://academic.oup.com/mbe/article-pdf/37/5/1530/33386032/msaa015.pdf</a>,
        <a href="https://doi.org/10.1093/molbev/msaa015">doi:10.1093/molbev/msaa015</a>.
    </dd>
    <dt>7</dt>
    <dd>Jacob&nbsp;L Steenwyk, III Buida, Thomas&nbsp;J, Abigail&nbsp;L Labella, Yuanning Li, Xing-Xing Shen, and
        Antonis Rokas.
        <span class="bibtex-protected">PhyKIT: a broadly applicable UNIX shell toolkit for processing and analyzing
            phylogenomic data</span>.
        <em>Bioinformatics</em>, 37(16):2325&ndash;2331, 02 2021.
        URL: <a
            href="https://doi.org/10.1093/bioinformatics/btab096">https://doi.org/10.1093/bioinformatics/btab096</a>, <a
            href="https://arxiv.org/abs/https://academic.oup.com/bioinformatics/article-pdf/37/16/2325/39948152/btab096.pdf">arXiv:https://academic.oup.com/bioinformatics/article-pdf/37/16/2325/39948152/btab096.pdf</a>,
        <a href="https://doi.org/10.1093/bioinformatics/btab096">doi:10.1093/bioinformatics/btab096</a>.
    </dd>
    <dt>8</dt>
    <dd>Jacob&nbsp;L. Steenwyk, Thomas&nbsp;J. Buida, Carla Gon<span class="bibtex-protected"><span
                class="bibtex-protected">ç</span></span>alves, Dayna&nbsp;C. Goltz, Grace Morales, Matthew&nbsp;E. Mead,
        Abigail&nbsp;L. LaBella, Christina&nbsp;M. Chavez, Jonathan&nbsp;E. Schmitz, Maria Hadjifrangiskou, Yuanning Li,
        and Antonis Rokas.
        <span class="bibtex-protected">BioKIT</span>: a versatile toolkit for processing and analyzing diverse types of
        sequence data.
        <em>biorxiv</em>, oct 2021.
        URL: <a href="https://doi.org/10.1101%2F2021.10.02.462868">https://doi.org/10.1101%2F2021.10.02.462868</a>, <a
            href="https://doi.org/10.1101/2021.10.02.462868">doi:10.1101/2021.10.02.462868</a>.
    </dd>
    <dt>9</dt>
    <dd>Jacob&nbsp;L. Steenwyk, Thomas&nbsp;J. Buida, III, Yuanning Li, Xing-Xing Shen, and Antonis Rokas.
        Clipkit: a multiple sequence alignment trimming software for accurate phylogenomic inference.
        <em>PLOS Biology</em>, 18(12):1&ndash;17, 12 2020.
        URL: <a href="https://doi.org/10.1371/journal.pbio.3001007">https://doi.org/10.1371/journal.pbio.3001007</a>, <a
            href="https://doi.org/10.1371/journal.pbio.3001007">doi:10.1371/journal.pbio.3001007</a>.
    </dd>
    <dt>10</dt>
    <dd>Chao Zhang, Maryam Rabiee, Erfan Sayyari, and Siavash Mirarab.
        Astral-iii: polynomial time species tree reconstruction from partially resolved gene trees.
        <em>BMC Bioinformatics</em>, 19(6):153, 2018.
        URL: <a href="https://doi.org/10.1186/s12859-018-2129-y">https://doi.org/10.1186/s12859-018-2129-y</a>, <a
            href="https://doi.org/10.1186/s12859-018-2129-y">doi:10.1186/s12859-018-2129-y</a>.
    </dd>
</dl>